### ISSUE TITLE
Marshalling: Generics

### ISSUE THREAD

**Issue by jasongin** [microsoft/node-api-dotnet#67]
JavaScript code should be able to:
  - Call generic methods (with explicit or _maybe_ implicit type arguments)
  - Get a property value or return value that is an instance of a .NET generic class, struct, or interface
  - Construct a new instance of a .NET generic class or struct (with explicit type arguments)
  - Get properties or invoke (non-generic) methods on an instance of a generic class, interface, or struct
  - Get static properties or invoke (non-generic) static methods on a generic type
  - Implement a .NET generic interface in a JavaScript class and pass an instance of it to .NET
  - Invoke or implement generic delegates

Note certain special-case generic types with mappings to built-in JavaScript types are implemented separately, including generic collections and generic tasks. This is about the general case.

In C#, type arguments can often be implicit, based on the argument types. Implicit type arguments often will not work well with JS since the type cannot as often be inferred from the JavaScript argument values. So JS code will usually need to specify explicit type arguments.

**Comment by jasongin** [comment]
## Projecting .NET Generics to JavaScript

Since the JavaScript runtime type system lacks generics, the projection of .NET generic APIs into
JavaScript needs to be considered carefully.

Note while TypeScript _does_ have generics, they are merely a compile-time facade and do not have
any effect on runtime binding or execution. So it is unfortunately not possible in most cases
to directly map .NET generics to TypeScript generics. (The exception is when mapping .NET generic
collections to TypeScript generic collections like `Array<T>` and `Map<TKey, TValue>`, which does
work.)

Following is a non-exhaustive list of scenarios for .NET generic APIs that the JavaScript interop
mechanism needs to support:

| Scenario | Example (C# syntax) |
|----------|---------------------|
| Calling a generic method                             | `Enum.Parse<T>()`
| Getting/setting a static property on a generic class | `Comparer<T>.Default`
| Calling a static method on a generic class           | `Comparer<T>.Create()`
| Creating an instance of a generic class              | `new KeyValuePair<TKey, TValue>()`
| Methods with both generic and non-generic overloads  | `Enum.Parse<T>()`<br/>`Enum.Parse()`
| Generic and non-generic classes with same name       | `TaskCompletionSource<T>`<br/>`TaskCompletionSource`

In general, JavaScript code needs a way to specify generic type arguments when working with .NET
generic types and generic methods. There are a few different syntax approaches that might be
considered for specifying type arguments...

### Rejected syntax (A)
A simple approach could be to specify the type argument(s) as the first parameter of a generic
method invocation. Consider the .NET `Enum.Parse<T>(string)` generic method that has a single
type argument and a string parameter. It might be invoked from JavaScript like this:
```JavaScript
// JavaScript
Enum.Parse(['System.DayOfWeek'], 'Tuesday');
```
But passing type arguments as a method parameter can't help with generic _types_.

### Rejected syntax (B)
To work consistently with generic types as well as generic methods, JavaScript code needs a way
to start with a _generic type or method definition_ by name, then specialize that to a generic
type or method by supplying type arguments. So logically the  generic type or method definition is
a function which takes type argument(s) and returns the specialized type or method:

```JavaScript
// JavaScript
Enum.Parse('System.DayOfWeek')('Tuesday');
Comparer('System.DateTime').Create();
```
While that syntax can work in many cases, there's still a problem if the type or method is available
in both generic and non-generic forms.

### Proposed syntax (C)
To resolve ambiguously generic type and method names it's necessary to project generic type and
method definitions to JavaScript with a special idenifier suffix. The dollar (`$`) is a good choice
because it is a valid identifier charater in JavaScript but not in C#, and because it looks like an
operator:

```JavaScript
// JavaScript
Enum.Parse$('System.DayOfWeek')('Tuesday');     // Call generic method
Comparer$('System.DateTime').Create()           // Call static method on generic class
new TaskCompletionSource$('System.DateTime')(); // Create instance of generic class
new TaskCompletionSource();                     // Create instance of non-generic class
```

### Type references
Example code above specifies generic type arguments using the full type name, e.g.
`'System.DateTime'`. But the JS marshaller can also support generic type arguments that are a
primitive type alias (`'int'`, `'string'`) or type objects. The type objects, as well as generic
type definition objects (the ones ending in `$`), can be obtained from the assembly object that
defines them, or directly from the `node-api-dotnet` module in the case of BCL types:
```JavaScript
// JavaScript
const dotnet = require('node-api-dotnet');
const DateTime = dotnet['System.DateTime'];
const KeyValuePair$ = dotnet['System.KeyValuePair$'];
const StringDatePair = KeyValuePair$('string', DateTime);
const pair = new StringDatePair('now', DateTime.Now);
```
As a convenience, it may be possible to directly request a fully-specified generic type using
angle-bracket syntax within the type name:
```JavaScript
// JavaScript
const StringDatePair = dotnet['System.KeyValuePair<string, System.DateTime>']
```

### Static binding / AOT
The above discussions assume dynamic binding, when the `node-api-dotnet` library can use
reflection to locate generic type and method definitions, specialize them, and invoke them.
But some of that is impossible to do in an ahead-of-time compiled environment. In a pure C#
module, the AOT compiler would be able to know exactly what type arguments are used with any
generic type or method, so it can generate specialized code accordingly. But dynamically
specifying generic type arguments from JavaScript would require reflection and code-generation,
whch are not supported in an AOT executable.

This means that C# Node API modules compiled as AOT cannot export generic types or methods.
They can still use generic types in properties or methods, when the type argument is specified
ahead of time. For example it is OK to use AOT to export a method that has a parameter of type
`KeyValuePair<string, int>`. But exporting a _generic_ method with type parameter `T` and method
parameter `KeyValuePair<string, T>` will not work with AOT.

**Comment by jasongin** [comment]
Update: The design proposed above is implemented in #142, with one significant difference: _string_ type references (either full names or aliases) are not supported. Instead, the types are also projected as objects in the namespace hierarchy.
